numba/core/retarget.py

```diff
@@ -0,0 +1,134 @@
+"""
+Implement utils for supporting retargeting of dispatchers.
+
+WARNING: Features defined in this file is experimental. The API may change
+         without notice.
+"""
+import abc
+import weakref
+
+from numba.core import errors
+
+
+class RetargetCache:
+    """Cache for retargeted dispatcher.
+
+    The cache uses the original dispatcher as the key.
+    """
+    container_type = weakref.WeakKeyDictionary
+
+    def __init__(self):
+        self._cache = self.container_type()
+        self._stat_hit = 0
+        self._stat_miss = 0
+
+    def save_cache(self, orig_disp, new_disp):
+        """Save a dispatcher associated with the given key.
+        """
+        self._cache[orig_disp] = new_disp
+
+    def load_cache(self, orig_disp):
+        """Load a dispatcher associated with the given key.
+        """
+        out = self._cache.get(orig_disp)
+        if out is None:
+            self._stat_miss += 1
+        else:
+            self._stat_hit += 1
+        return out
+
+    def items(self):
+        """Returns the contents of the cache.
+        """
+        return self._cache.items()
+
+    def stats(self):
+        """Returns stats regarding cache hit/miss.
+        """
+        return {'hit': self._stat_hit, 'miss': self._stat_miss}
+
+
+class BaseRetarget(abc.ABC):
+    """Abstract base class for retargeting logic.
+    """
+    @abc.abstractmethod
+    def check_compatible(self, orig_disp):
+        """Check that the retarget is compatible.
+
+        This method does not return anything meaningful (e.g. None)
+        Incompatibility is signalled via raising an exception.
+        """
+        pass
+
+    @abc.abstractmethod
+    def retarget(self, orig_disp):
+        """Retarget the givern dispatcher and returns a new dispatcher-like
+        callable. Or, return the original dispatcher if the the target_backend
+        will not change.
+        """
+        pass
+
+
+class BasicRetarget(BaseRetarget):
```

Comment:
This isn't tested?

Reply:
It is tested in `numba/tests/test_retargeting.py` because `CustomCPURetarget` implements `BasicRetarget`

Reply:
Apologies, the above wasn't clear. I was thinking about the hits/misses/stats part to ensure the cache is behaving correctly.